pkg/check/table_structure.go

```diff
@@ -77,73 +149,249 @@ func (c *TablesChecker) Name() string {
-			return errors.Trace(err)
+		option := c.checkColumnDef(def)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
 	// check constrains
 	for _, cst := range st.Constraints {
-		err = c.checkConstraint(cst)
-		if err != nil {
-			return errors.Trace(err)
+		option := c.checkConstraint(cst)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
 
 	// check options
 	for _, opt := range st.Options {
-		err = c.checkTableOption(opt)
-		if err != nil {
-			return errors.Trace(err)
+		option := c.checkTableOption(opt)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
-	return nil
+	return options
 }
 
-func (c *TablesChecker) checkColumnDef(def *ast.ColumnDef) error {
+func (c *TablesChecker) checkColumnDef(def *ast.ColumnDef) *incompatibilityOption {
 	return nil
 }
 
-func (c *TablesChecker) checkConstraint(cst *ast.Constraint) error {
+func (c *TablesChecker) checkConstraint(cst *ast.Constraint) *incompatibilityOption {
 	switch cst.Tp {
 	case ast.ConstraintForeignKey:
-		c.warnings++
-		return errors.Errorf("Foreign Key is parsed but ignored by TiDB.")
+		return &incompatibilityOption{
+			state:       StateWarning,
+			instruction: fmt.Sprintf("please ref document: https://github.com/pingcap/docs-cn/blob/master/sql/ddl.md"),
+			errMessage:  fmt.Sprintf("Foreign Key %s is parsed but ignored by TiDB.", cst.Name),
+		}
 	}
+
 	return nil
 }
 
-func (c *TablesChecker) checkTableOption(opt *ast.TableOption) error {
+func (c *TablesChecker) checkTableOption(opt *ast.TableOption) *incompatibilityOption {
 	switch opt.Tp {
 	case ast.TableOptionCharset:
 		// Check charset
 		cs := strings.ToLower(opt.StrValue)
 		if cs != "binary" && !charset.ValidCharsetAndCollation(cs, "") {
-			c.errors++
-			return errors.Errorf("Unsupported charset %s", opt.StrValue)
+			return &incompatibilityOption{
+				state:       StateFailure,
+				instruction: fmt.Sprintf("please ref document: https://github.com/pingcap/docs-cn/blob/master/sql/character-set-support.md"),
+				errMessage:  fmt.Sprintf("unsupport charset %s", opt.StrValue),
+			}
 		}
 	}
 	return nil
 }
+
+// ShardingTablesCheck checks consistency of sharding table structures
+// * check whether they have same column list
+// * check whether they have auto_increment key
+type ShardingTablesCheck struct {
+	dbs    map[string]*sql.DB
+	tables map[string]map[string][]string // instance => {schema: [table1, table2, ...]}
+}
+
+// NewShardingTablesCheck returns a Checker
+func NewShardingTablesCheck(dbs map[string]*sql.DB, tables map[string]map[string][]string) Checker {
+	return &ShardingTablesCheck{
+		dbs:    dbs,
+		tables: tables,
+	}
+}
+
+// Check implements Checker interface
+func (c *ShardingTablesCheck) Check(ctx context.Context) *Result {
+	r := &Result{
+		Name:  c.Name(),
+		Desc:  "check consistency of sharding table structures",
+		State: StateSuccess,
+	}
+
+	var (
+		hasAutoIncrementKeyTableNum int
+		hasAutoIncrementKeyTables   []string
```

Comment:
only use `hasAutoIncrementKeyTables` is enough ?

Reply:
it should be `map[string][]string`